mc_calculator/recipe_logic.py

```diff
@@ -73,71 +91,123 @@ def create_recipe():
+        desired_quantity (int): The desired quantity of the final product.
+
+    Returns:
+        dict: A dictionary of ingredients and their required quantities.
+        list: A list of steps involved in making the recipe.
+    """
+    ingredients_needed = {}
+    steps = []
+
+    desired_runs = math.ceil(desired_quantity / recipe.output_count)
+    ingredients_needed.update(calculate_single_recipe_ingredients(recipe, desired_runs))
+
+    for nested_id, quantity_needed in recipe.nested_recipes.items():
+        nested_recipe = db.fetch_recipe_by_id(nested_id)
+        if nested_recipe:
+            (
+                nested_ings,
+                nested_steps,
+                runs_needed,
+            ) = calculate_nested_recipe_ingredients(
+                nested_recipe, quantity_needed, desired_quantity
+            )
+            steps.append(
+                (
+                    nested_recipe.name,
+                    runs_needed,
+                    nested_recipe.output_count,
+                    nested_steps,
+                )
+            )
+            for ing, qty in nested_ings.items():
+                ingredients_needed[ing] = ingredients_needed.get(ing, 0) + qty
+
+    return ingredients_needed, steps
 
-    This function takes a recipe name and the desired quantity of the final product, then calculates
-    the total amount of each ingredient needed. It handles both simple and nested recipes.
+
+def calculate_single_recipe_ingredients(recipe, desired_runs):
+    """
+    Calculates the ingredients required for a given recipe based on the number of desired runs.
 
     Args:
-        recipe_name (str): The name of the recipe for which ingredients are to be calculated.
+        recipe (Recipe): The recipe for which ingredients are needed.
+        desired_runs (int): The number of times the recipe needs to be executed.
+
+    Returns:
+        dict: A dictionary of ingredients and their required quantities.
+    """
+    ingredients_needed = {}
+    for ingredient, quantity in recipe.ingredients.items():
+        total_quantity = quantity * desired_runs
+        ingredients_needed[ingredient] = (
+            ingredients_needed.get(ingredient, 0) + total_quantity
+        )
+    return ingredients_needed
+
+
+def calculate_nested_recipe_ingredients(
+    nested_recipe, quantity_needed, desired_quantity
+):
+    """
+    Calculates the ingredients required for nested recipes.
+
+    Args:
+        nested_recipe (Recipe): The nested recipe.
+        quantity_needed (int): The quantity of the nested recipe required.
         desired_quantity (int): The desired quantity of the final product.
 
     Returns:
-        None: This function prints the required ingredients and their quantities to the console.
+        dict: A dictionary of ingredients and their quantities for the nested recipe.
+        list: A list of steps involved in the nested recipe.
+        int: The number of runs needed for the nested recipe.
```

Comment:
Looks like you're using a google docstring. For this type of return you'll probably want to describe the return as tuple of objects. Google Docstring style guide has some pointers on this:

https://google.github.io/styleguide/pyguide.html#doc-function-returns

They even have specific guidance on tuple-type syntax descriptions here:

https://google.github.io/styleguide/pyguide.html#3199-tuples-vs-lists

So maybe would look like:

```
Returns:

    tuple[dict, list, int]: A tuple containing the dictionary of ingredients/quantities, a list of steps, and the number of runs needed for a nested recipe
```

Reply:
Thanks for guidance, I'll implement this change immediately. The resources really helped.